NGSTACK-977 tags visibility

AntePrkacin opened the pull request:
Added `isHidden` and `isInvisible` fields to Tag objects. Updated the database schema and updated the whole workflow for hiding and unhiding a tag (Service, Mapper, Handler, Gateway, ...).

Added columns 'Hidden' and 'Invisible' to table that show children tags of some tag in admin interface. 
In admin interface added button for hiding a tag (or if a tag is already hidden, then this button will unhide it). 
Also added buttons 'Hide selected tags' and 'Unhide selected tags' for the checked children tags. 

When a tag is hidden, the route `/tags/view/Topics/{tag_name}` will throw 404 and return information that 'Tag is hidden' or 'Tag is hidden by parent'. 

If a tag is hidden, then the user won't be able to add that tag to some 'eztags' content field for some content.
If some content already has a hidden tag, then this tag will have ' (hidden)' appended to it and it will be in gray color.

AntePrkacin committed:
NGSTACK-977 add 'isHidden' and 'isInvisible' properties to Tag object in both API and persistence layer

AntePrkacin committed:
NGSTACK-977 update database schema to include 'is_hidden' and 'is_invisible' columns in 'eztags' table

AntePrkacin committed:
NGSTACK-977 add support for 'isHidden' and 'isInvisible' properties in DoctrineDatabase, TagsMapper and Mapper classes

AntePrkacin committed:
NGSTACK-977 add support for 'isHidden' and 'isInvisible' fields in translation files and display them in admin UI

AntePrkacin committed:
NGSTACK-977 add 'Hide tag' button and its functionality in admin interface

AntePrkacin committed:
NGSTACK-977 add 'Unhide tag' button and its functionality in admin interface

AntePrkacin committed:
NGSTACK-977 fix bug when creating queries and update translations to show when a tag is hidden by parent

AntePrkacin committed:
NGSTACK-977 add functionality for hiding and unhiding selected children tags

AntePrkacin committed:
NGSTACK-977 throw 404 on frontend if tag is hidden

AntePrkacin committed:
NGSTACK-977 add 'disabled' property to hidden tags and disable adding hidden tags to content

AntePrkacin committed:
NGSTACK-977 add 'hidden' text after hidden tags in eztags content field

emodric commented on bundle/API/Repository/TagsService.php:
```diff
@@ -286,6 +286,22 @@ public function newSynonymCreateStruct(int $mainTagId, string $mainLanguageCode)
+    /**
+     * Hides $tag.
+     *
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\UnauthorizedException If the current user is not allowed to delete this tag
```
Wrong exception text `If the current user is not allowed to delete this tag`. Shoud be hide instead of delete.

Same goes for unhide.

AntePrkacin replied:
Fixed PHPDocs: https://github.com/netgen/TagsBundle/pull/169/commits/c698f263fc232e8ac50a350b7362d9f5ccca6bcc

Will probably rename 'unhide' to 'reveal', as Petar suggested below

emodric commented on bundle/Controller/TagViewController.php:
```diff
@@ -14,6 +14,14 @@ final class TagViewController extends Controller
     public function viewAction(TagView $view): TagView
     {
+        if ($view->getTag()->isHidden) {
+            throw $this->createNotFoundException('Tag is hidden.');
```
We don't need the message here. As this is the frontend, we do not want to communicate any revealing info.

As far as the user is concerned, the tag does not exist and that's it.

Also, here, you can only check for visible/invisible since when the tag is hidden, it should automatically be invisible too. That is, it is not possible for the tag to be hidden but not invisible too.

AntePrkacin replied:
Removed the message: https://github.com/netgen/TagsBundle/pull/169/commits/c698f263fc232e8ac50a350b7362d9f5ccca6bcc

emodric commented on bundle/Core/Event/TagsService.php:
```diff
@@ -226,6 +226,16 @@ public function newTagUpdateStruct(): TagUpdateStruct
 
+    public function hideTag(Tag $tag): void
+    {
+        $this->service->hideTag($tag);
```
No event dispatching here and in unhide tag method?

AntePrkacin replied:
Forgot about them... added them now: https://github.com/netgen/TagsBundle/pull/169/commits/69672a56b8c4b217246826da40559ecd4fe7a13c

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -844,6 +844,62 @@ public function deleteTag(int $tagId): void
+                $query->expr()->like('path_string', ':path_string'),
+            )
+            ->andWhere(
+                $query->expr()->neq('id', ':tag_id'),
```
When hiding a tag, the hidden root tag should also be invisible, so we don't need this `neq` condition.

This is the way Ibexa works when hiding content, so there is no need to do this differently here. It will only cause confusion in the long run.

AntePrkacin replied:
Ok, didn't know that this was how Ibexa works. 
Removed the `neq`: https://github.com/netgen/TagsBundle/pull/169/commits/f8bcf8d7540a357d4899b6cf99066d04e54e1e6e

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -844,6 +844,62 @@ public function deleteTag(int $tagId): void
+                $query->expr()->like('path_string', ':path_string'),
+            )
+            ->andWhere(
+                $query->expr()->neq('id', ':tag_id'),
```
Same remark about root tag invisibility goes here.

Also, I think we need to consider here what happens when some tag is already hidden explicitly below the tag being unhidden here. In that case, we can't really make tags below the already hidden tag invisible to maintain data consistency.  We should see how Ibexa does this and just copy the implementation from there.

AntePrkacin replied:
Fixed the tag invisibility in the above comment. 

Regarding the problem you mentioned about looking at the descendant tags and leaving them invisible if they have at least one parent that is explicitly hidden, I fixed it here: 
https://github.com/netgen/TagsBundle/pull/169/commits/8b20a727def5701cca7d8009ab25b8b6113facee https://github.com/netgen/TagsBundle/pull/169/commits/54ec1e9e6de56c8e324578e794b5285a762bd07b https://github.com/netgen/TagsBundle/pull/169/commits/45b2f03e1bd231af43ab0d233dc96f23e51c58a6

emodric commented on bundle/Core/Persistence/Legacy/Tags/Mapper.php:
```diff
@@ -35,6 +35,8 @@ public function createTagInfoFromRow(array $row): TagInfo
         $tagInfo->alwaysAvailable = (bool) ((int) $row['language_mask'] & 1);
         $tagInfo->mainLanguageCode = $this->languageHandler->load($row['main_language_id'])->languageCode;
         $tagInfo->languageIds = $this->languageMaskGenerator->extractLanguageIdsFromMask((int) $row['language_mask']);
+        $tagInfo->isHidden = (bool) ((int) $row['is_hidden']);
```
We don't need double cast here. You probably copied the thing from `(bool) ((int) $row['language_mask'] & 1)`, but that has double cast due to bitwise and operator.

AntePrkacin replied:
Removed the double cast, kept only the `bool` one: https://github.com/netgen/TagsBundle/pull/169/commits/e5b2f5f16e989715fe5a87042db88dcd9780effb

emodric commented on bundle/Resources/config/policies.yaml:
```diff
@@ -9,3 +9,5 @@ tags:
     deletesynonym: ~
     makesynonym: ~
     merge: ~
+    hide: ~
```
You're missing `hidesysnonym` and `unhidesynonym` policies here. Also do we really need policies for both hide and unhide?

AntePrkacin replied:
Added `hidesynonym` and `unhidesynonym` policies: https://github.com/netgen/TagsBundle/pull/169/commits/7e88bdde7cd76a13bee70f33be1009e27885f088

I also thought that just one policy for hiding tags is enough, but I made two just for clarity. 
If you want, I can only leave 'hide' and 'hidesynonym' policies, which will mean that the user can both hide and reveal tags.

AntePrkacin replied:
@emodric any changes about this? Should I just make one policy (e.g. visibility) or leave it as it is now?

emodric replied:
I think one is enough. I don't see a usecase where someone can only hide, but not reveal and vice versa.

emodric replied:
I would also remove the "synonym" variant too, which would leave only one `hide` policy for everything.

AntePrkacin replied:
Left only the 'hide' policy, removed the rest: https://github.com/netgen/TagsBundle/pull/169/commits/71a91025f867df4c18495736f8cb022f0ffacfbf

emodric commented on bundle/Resources/views/admin/eztags_content_field.html.twig:
```diff
@@ -1,5 +1,15 @@
+        >
+            {{ tag.keyword }}
+            {% if tag.isHidden %}
+                (hidden)
```
Needs translations too.

AntePrkacin replied:
Added translations here for '(hidden)' and '(hidden by parent)': https://github.com/netgen/TagsBundle/pull/169/commits/dece52cad1c85d0ca6adca828b40db5fbe8cbf16

AntePrkacin committed:
NGSTACK-977 fix php-cs-fixer and phpstan reports

pspanja commented on bundle/API/Repository/TagsService.php:
```diff
@@ -286,6 +286,22 @@ public function newSynonymCreateStruct(int $mainTagId, string $mainLanguageCode)
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\UnauthorizedException If the current user is not allowed to delete this tag
+     * @throws \Ibexa\Contracts\Core\Repository\Exceptions\NotFoundException If the specified tag is not found
+     */
+    public function unhideTag(Tag $tag): void;
```
I would go with `revealTag()` here, to be inline with the newer implementation in Repository API and with labeling in the Admin UI. (We do have `LocationService::unhideLocation()`, but corresponding newer implementation in `ContentService` is named `revealContent()`.

AntePrkacin replied:
Changed all 'unhide' text to 'reveal' regarding the tags: https://github.com/netgen/TagsBundle/pull/169/commits/543afc4bebad2f73827f441de210c0464723802f

AntePrkacin committed:
NGSTACK-977 add 'is_hidden' and 'is_invisible' fields to test database schemas and fix MapperTest

AntePrkacin committed:
NGSTACK-977 fix TagsIntegrationTest by adding 'isHidden' and 'isInvisible' fields

AntePrkacin committed:
NGSTACK-977 update phpdocs and remove tag visibility information being shown to the user

AntePrkacin committed:
NGSTACK-977 modify gateway to make the hidden tag also invisible

AntePrkacin committed:
NGSTACK-977 remove unnecessary cast to int for 'is_hidden' and 'is_invisible' fields in Mapper

AntePrkacin committed:
NGSTACK-977 add 'hidesynonym' and 'unhidesynonym' policies

AntePrkacin committed:
NGSTACK-977 add translation domain and make tag visibility info translatable in 'eztags' content field

AntePrkacin committed:
NGSTACK-977 change all 'unhide' text to 'reveal' regarding the tags visibility

AntePrkacin committed:
NGSTACK-977 make descendant tags who have one hidden parent stay invisible when revealing a tag

AntePrkacin committed:
NGSTACK-977 fix logic for revealing a hidden tag

AntePrkacin committed:
NGSTACK-977 set 'is_hidden' and 'is_invisible' fields correctly when creating a tag

AntePrkacin committed:
NGSTACK-977 update logic for revealing a tag to handle both ancestor and descendant hidden tags

AntePrkacin committed:
NGSTACK-977 fix a test for creating a tag by adding 'is_invisible' property

AntePrkacin committed:
NGSTACK-977 add events for hiding and revealing a tag and add tests for it

AntePrkacin committed:
NGSTACK-977 add 'show_hidden' parameter in the configTreeBuilder and contextualized it

AntePrkacin committed:
NGSTACK-977 inject configResolver into SiteAccessAware/TagsService, add 'show_hidden' argument to necessary functions in the TagsService implementations

AntePrkacin committed:
NGSTACK-977 add 'show_hidden' argument to necessary functions in the SPI/Persistence layer and pass it on down to the Gateway

AntePrkacin committed:
NGSTACK-977 add 'show_hidden' argument to necessary functions in the Gateway class and update the logic to handle when 'showHidden' config param is false

AntePrkacin commented:
Added `show_hidden` config parameter to the configTreeBuilder. By default it is set to true. This parameter can be changed for each site access. 

Implemented the following methods of the `TagsService` class to look at this parameter:
- loadTagChildren()
- getTagChildrenCount()
- loadTagsByKeyword()
- getTagsByKeywordCount()
- searchTags()
- loadTagSynonyms()
- getTagSynonymCount()

https://github.com/netgen/TagsBundle/pull/169/commits/086e95491e474b4618c1844ff39be4078fff076b https://github.com/netgen/TagsBundle/pull/169/commits/4af3ecd15db43be123c07531a1b36306763395ab https://github.com/netgen/TagsBundle/pull/169/commits/0f8c7dd4b1233019e9bb0c8e9e54c26aab94dd3c https://github.com/netgen/TagsBundle/pull/169/commits/75858b832236bda673435fed89fb52f88972cb29

AntePrkacin committed:
NGSTACK-977 update PHPDoc for hide and reveal methods and add ' (hidden)' text next to synonym name if the synonym is hidden

AntePrkacin committed:
NGSTACK-977 add ' (hidden)' next to tag keyword in the TagTreeData if the tag is hidden

AntePrkacin committed:
NGSTACK-977 update DoctrineDatabaseTest and TagsHandlerTest with tests for revealTag() and hideTag() methods

AntePrkacin committed:
NGSTACK-977 add 'hide_tag' and 'reveal_tag' data to the TagTree in TreeController

AntePrkacin committed:
NGSTACK-977 add visibility information for children when listing them for some tag

AntePrkacin committed:
NGSTACK-977 not show hidden tags when trying to search for them in eztags field

AntePrkacin committed:
NGSTACK-977 update postgresql schema and legacy.yaml file with 'is_hidden' and 'is_invisible' properties

pspanja commented on bundle/API/Repository/Values/Tags/Tag.php:
```diff
@@ -104,6 +106,16 @@ final class Tag extends ValueObject
+    protected bool $isHidden;
+
+    /**
+     * Indicates if the Tag object is located under another hidden Tag object.
```
Rewrite:

> Indicates that the Tag object is not visible, being either hidden itself, or implicitly hidden by parent or ancestor Tag object.

AntePrkacin replied:
Updated - https://github.com/netgen/TagsBundle/pull/169/commits/81d40076cf9a82826ebab7676d640243fbfa4396

pspanja commented on bundle/API/Repository/Values/Tags/Tag.php:
```diff
@@ -104,6 +106,16 @@ final class Tag extends ValueObject
     protected ?string $prioritizedLanguageCode;
 
+    /**
+     * Indicates if the Tag object is visible or not.
```
Rewrite:

> Indicates that the Tag is hidden.

AntePrkacin replied:
Updated - https://github.com/netgen/TagsBundle/pull/169/commits/81d40076cf9a82826ebab7676d640243fbfa4396

pspanja commented on bundle/Controller/Admin/FieldController.php:
```diff
@@ -37,6 +37,7 @@ public function autoCompleteAction(Request $request): JsonResponse
         $searchResult = $this->tagsService->searchTags(
             $request->query->get('searchString') ?? '',
             $request->query->get('locale') ?? '',
+            showHidden: false,
```
Should this be controlled through the configuration as well?

AntePrkacin replied:
I hardcoded here _false_ because the `autoCompleteAction()` method is used for searching a tag to add it in `eztags` field. If the **showHidden** parameter wasn't _false_ here, then the user could add a hidden tag to some content by searching for it when editing a content.

pspanja replied:
True, but there might be a use case where you prepare the Tags and Content in advance, and just reveal the Tags at some time point. I'll leave it to @emodric.

cc @Ljudevit for info, maybe you have a use case for this?

Ljudevit replied:
I would like to agree with @AntePrkacin on this one that we should always avoid outputting hidden tags in the autocomplete actions like searching for tags in tag fields, but there can be some quite silly edge-cases where editor would hide the tag so tags/view page is not visible yet, but they would like to start filling in the content with those hidden tags before revealing the tag all-together.

For sure this configuration parameter must be set to false as default, but for those pesky edge-cases, a parameter is needed.

emodric replied:
Yes, parameter is definitely needed.

Also, please do not use named parameters, at least not mixed like this, it just looks wrong :D Either change all of them to use named parameters, or none.

AntePrkacin replied:
Added config parameter called `autocomplete_provide_hidden_tags`. Set it to `false` by default. Didn't use named parameter: https://github.com/netgen/TagsBundle/pull/169/commits/535890e6d374eabd5ed6416213d2098ccb3c5471 https://github.com/netgen/TagsBundle/pull/169/commits/96279be1f19e95ea907223050479cd429fc5fa78

Also, just to follow up on this discussion - if some tag is hidden by parent (so it is invisible, but not hidden), should the autocomplete method show this tag or not? Probably not, right?

emodric replied:
Hidden and invisible is, for all intents and purposes, basically the same. The difference comes from hiding/unhiding procedure, since you need to care about whether one tag was hidden automatically or manually.

AntePrkacin replied:
Ok, thanks!
Updated the autocomplete process in searching to not show tags that are invisible (hidden by parent): https://github.com/netgen/TagsBundle/pull/169/commits/9a00e4ed9ad5bba7c03fcf6295eb273572d6b0f4

pspanja commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -500,9 +500,49 @@ public function childrenAction(Request $request, ?Tag $tag = null): Response
+    {
+        $this->denyAccessUnlessGranted('ibexa:tags:hide' . ($tag->isSynonym() ? 'synonym' : ''));
+
+        $this->tagsService->hideTag($tag);
```
Do we need to check the submitted button and CSRF token here?

emodric replied:
Yes, we do. There is an example in `deleteTagAction` method.

AntePrkacin replied:
Added CSRF token check to `hideAction()` method in TagController (https://github.com/netgen/TagsBundle/pull/169/commits/550f0e3b3f294a7ca8633ca3d9905264247eb4ce)

pspanja commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -500,9 +500,49 @@ public function childrenAction(Request $request, ?Tag $tag = null): Response
+    {
+        $this->denyAccessUnlessGranted('ibexa:tags:reveal' . ($tag->isSynonym() ? 'synonym' : ''));
+
+        $this->tagsService->revealTag($tag);
```
Do we need to check the submitted button and CSRF token here?

emodric replied:
Yes, we do. There is an example in `deleteTagAction` method.

AntePrkacin replied:
Added CSRF token check to `revealAction()` method in TagController (https://github.com/netgen/TagsBundle/pull/169/commits/550f0e3b3f294a7ca8633ca3d9905264247eb4ce)

pspanja commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -666,6 +706,60 @@ public function deleteTagsAction(Request $request, ?Tag $parentTag = null): Resp
+            $this->tagsService->revealTag($tagObject);
+        }
+
+        $this->addFlashMessage('success', 'tags_reveal');
```
`tags_revealed`*

AntePrkacin replied:
Updated to `tags_revealed` - https://github.com/netgen/TagsBundle/pull/169/commits/a720391afb7e8ebaff55da0defe577abe07c88d9

pspanja commented on bundle/Controller/Admin/TreeController.php:
```diff
@@ -121,11 +126,25 @@ private function getTagTreeData(Tag $tag, bool $isRoot = false): array
     {
         $synonymCount = $this->tagsService->getTagSynonymCount($tag);
 
+        $text = $this->escape($tag->keyword);
```
Generating the name/text now has enough logic that extracting it into a method seems like a good idea.

AntePrkacin replied:
Extracted the logic for formatting the TagTreeText into a separate function - https://github.com/netgen/TagsBundle/pull/169/commits/a69c38cf12cab6db5c58c5070ff7c6e88e74ef11

pspanja commented on bundle/API/Repository/TagsService.php:
```diff
@@ -76,7 +76,7 @@ public function loadTagByUrl(string $url, array $languages): Tag;
      * @return \Netgen\TagsBundle\API\Repository\Values\Tags\TagList
      */
-    public function loadTagChildren(?Tag $tag = null, int $offset = 0, int $limit = -1, ?array $languages = null, bool $useAlwaysAvailable = true): TagList;
+    public function loadTagChildren(?Tag $tag = null, int $offset = 0, int $limit = -1, ?array $languages = null, bool $useAlwaysAvailable = true, ?bool $showHidden = null): TagList;
```
Possibly we should have a default value for `$showHidden` here, instead of resolving it in the Persistence implementation. @emodric I'll leave that up to you :)

(Then the same applies for the similar cases below)

Edit: ignore the above, I think the parameter should not be optional in the Persistence layer, we probably need it to be optional here so we can properly handle it in the siteaccess-aware implementation.

emodric replied:
I agree with @pspanja , however, nullable bool seems weird. Either we show or hide hidden tags, there's no third option for `null`. So `bool $showHidden = false` is probably good enough.

pspanja replied:
@emodric one question though, if here the default is not `null`, how do we detect explicitly passed value in the siteaccess-aware layer?

emodric replied:
Hm. Good point. Although, do we even have the need for explicit API layer control of this flag? Can we rely only on siteaccess aware config?

pspanja replied:
I would say we do, this is not only for rendering, it might be used in import for example.

emodric replied:
Import scripts can be run in admin siteaccess if needed.

It's just weird design, is all, since API interface presumes that somwhere down below, the value we give to the method might be overriden.

But okay, if there are valid usecases, I can live with it.

AntePrkacin replied:
> If `$showHidden` were not nullable, as mentioned before, you wouldn't need double check here and in other places.

Do I then leave it to be null by default or should I change it?

emodric replied:
I think this is okay, but let @pspanja confirm too.

pspanja replied:
In general I'm OK with it, but I'm not clear about this:

> It's just weird design, is all, since API interface presumes that somwhere down below, the value we give to the method might be overriden.

That should definitely not be the case, if the value was passed to the API method call, it must not be overridden somewhere below. That's why in the original comment I suggested making the parameter mandatory in the Persistence layer. It should be controlled from API layer only - either through siteaccess aware layer or by being explicitly given when calling the method.

emodric replied:
What I meant was, it looks like API interface is assuming the implementation, or at least one of them.

In terms of API, we either show hidden tags, or don't show hidden tags, there's no third option. And then along comes the third option, that says: "let the implementation decide", which is kinda ambiguous since the user does not know what the end result will be. That's why I'm saying that this API interface is looking weird.

But no matter, we already decided that this is relatively fine, so lets just move on :)

pspanja replied:
As mentioned f2f, we need to document that `null` is interpreted `true` (or `false`), and then it's kind of OK :)

pspanja commented on bundle/Core/Persistence/Legacy/Tags/Mapper.php:
```diff
@@ -60,6 +62,8 @@ public function extractTagListFromRows(array $rows): array
                 $tag->alwaysAvailable = (bool) ((int) $row['language_mask'] & 1);
                 $tag->mainLanguageCode = $this->languageHandler->load($row['main_language_id'])->languageCode;
                 $tag->languageIds = $this->languageMaskGenerator->extractLanguageIdsFromMask((int) $row['language_mask']);
+                $tag->isHidden = (bool) ((int) $row['is_hidden']);
```
Double case should not be needed here as well.

AntePrkacin replied:
Removed the double cast - https://github.com/netgen/TagsBundle/pull/169/commits/a69c38cf12cab6db5c58c5070ff7c6e88e74ef11

AntePrkacin committed:
NGSTACK-977 update PHPDoc for isHidden and isInvisible Tag properties

AntePrkacin committed:
NGSTACK-977 modify 'tags_reveal' to 'tags_revealed' translation message

AntePrkacin committed:
NGSTACK-977 extract TagTreeText formatting into a separate function and remove double cast from the Mapper

pspanja commented:
@AntePrkacin you can add upgrade instructions in `doc/UPGRADE.md`, including database migration, as was already done here: https://github.com/netgen/TagsBundle/blob/master/doc/UPGRADE.md#notes-on-upgrading-to-ez-platform-22

Use `5.4.0` as the next version.

AntePrkacin committed:
NGSTACK-977 add upgrade instructions for tag visibility to UPGRADE.md file

AntePrkacin commented:
Updated the `doc/UPGRADE.md` file with update instructions and sql queries to include `is_hidden` and `is_invisible` columns (https://github.com/netgen/TagsBundle/pull/169/commits/08ef43df4dd3ce63849626471e3af1f7834c845e). Correct me if anything needs to be fixed.

emodric commented on bundle/Controller/Admin/TagController.php:
```diff
@@ -666,6 +706,60 @@ public function deleteTagsAction(Request $request, ?Tag $parentTag = null): Resp
         );
     }
 
+    public function hideTagsAction(Request $request, ?Tag $parentTag = null): Response
```
Same here with CSRF. What we need to do here is create a Symfony multiselect form for selecting the tags in lieu of `moveTagsAction` method. Same goes for `revealTagsAction`.

AntePrkacin replied:
@emodric OK. Just to confirm that I understood correctly - I will remove the 'Hide selected tags' and 'Reveal selected tags' buttons below the children list - `childrenAction()` method. Instead, I will put two new buttons (somewhere else) that tell 'Hide kids' or 'Reveal kids'. When either of them is pressed, a multiselect form will show up where the user will be able to select the desired tags they want to hide or reveal.

emodric replied:
No no, do it in the same way as it is now, but with Symfony form in lieu of "move tags". Try and see if you can reuse the existing form type. It's now called `MoveTagsType`, try to reuse it by renaming it to e.g. `MultiselectTagsType` or similar.

AntePrkacin replied:
Renamed `MoveTagsType` form to `MultiselectTagsType` form and used it when hiding and revealing multiple tags:
https://github.com/netgen/TagsBundle/pull/169/commits/c05ab163f670f31f974c99108bf7040c704adc15 https://github.com/netgen/TagsBundle/pull/169/commits/07866bc4feadfbbf5d17788e975d2ba06a68826f

emodric replied:
@AntePrkacin Sounds good :+1:

emodric commented on bundle/Controller/Admin/TreeController.php:
```diff
@@ -165,13 +170,37 @@ private function getTagTreeData(Tag $tag, bool $isRoot = false): array
     }
 
-    private function escape(string $string): string
+    private function formatTagTreeText(Tag $tag): string
```
We shouldn't mix escaping and formating. That is, leave `escape` method as is, and just reuse is in the new `formatTagTreeText` method. After all, all of the text should be escaped AFTER formatting it, and not just the starting text.

AntePrkacin replied:
Separated the _escape_ logic from the _formatting_ logic: https://github.com/netgen/TagsBundle/pull/169/commits/4a174cc82497a6c74cee527b8a6100cbfb53ed94

emodric commented on bundle/Core/Persistence/Legacy/Tags/Gateway/DoctrineDatabase.php:
```diff
@@ -154,6 +161,13 @@ public function getChildren(int $tagId, int $offset = 0, int $limit = -1, ?array
 
         $query = $this->createTagFindQuery($translations, $useAlwaysAvailable);
+
+        if ($showHidden !== null && $showHidden === false) {
```
If `$showHidden` were not nullable, as mentioned before, you wouldn't need double check here and in other places.

emodric commented on bundle/DependencyInjection/Configuration.php:
```diff
@@ -155,6 +155,10 @@ public function getConfigTreeBuilder(): TreeBuilder
                     ->end()
                 ->end()
+            ->end()
+            ->booleanNode('show_hidden')
```
We should be explicit and name this `show_hidden_tags`.

AntePrkacin replied:
Renamed `show_hidden` to `show_hidden_tags` here in the ConfigTree as well as in Service, Handler and Gateway layers (renamed parameter `$showHidden` to `$showHiddenTags`): https://github.com/netgen/TagsBundle/pull/169/commits/6692248b769f5df03decc1f4f343c1f8cc419b89

emodric commented on bundle/Resources/sql/mysql/schema.sql:
```diff
@@ -9,6 +9,8 @@ CREATE TABLE `eztags` (
   `remote_id` varchar(100) NOT NULL default '',
   `main_language_id` int(11) NOT NULL default '0',
   `language_mask` int(11) NOT NULL default '0',
+  `is_hidden` tinyint NOT NULL DEFAULT '0',
```
We also need MySQL and PostgreSQL upgrade scripts (see `bundle/Resources/sql/upgrade` folder).

AntePrkacin replied:
Added MySQL and PostgreSQL upgrade scripts: https://github.com/netgen/TagsBundle/pull/169/commits/203dcd68ec7bc02cbb48580491eae3c3020f86ac

emodric commented on doc/UPGRADE.md:
````diff
@@ -1,6 +1,28 @@
+If you're using MySQL, run the commands below:
+
+```sql
+ALTER TABLE `eztags` ADD COLUMN `is_hidden` TINYINT NOT NULL DEFAULT 0;
````
These should go to an upgrade SQL script as mentioned before and just referenced here.

AntePrkacin replied:
Added MySQL and PostgreSQL upgrade scripts and referenced them here (`doc/UPGRADE.md`): https://github.com/netgen/TagsBundle/pull/169/commits/203dcd68ec7bc02cbb48580491eae3c3020f86ac

AntePrkacin committed:
NGSTACK-977 update hideAction() and revealAction() methods for hiding and revealing a tag with CSRF token check

AntePrkacin committed:
NGSTACK-977 separate escape logic from the formatting logic for tag keyword in TreeController

AntePrkacin committed:
NGSTACK-977 rename 'showHidden' to 'showHiddenTags' parameter in configTree, Service, Handler and Gateway layers

AntePrkacin committed:
NGSTACK-977 add MySQL and PostgreSQL upgrade scripts for 'is_hidden' and 'is_invisible' columns

AntePrkacin committed:
NGSTACK-977 fix a call to searchTags() method in FieldController

AntePrkacin committed:
NGSTACK-977 rename MoveTagsType form to MultiselectTagsType form and add 'show_parent_field' option to the form

AntePrkacin committed:
NGSTACK-977 update hideTagsAction and revealTagsAction methods in TagController to use MultiselectTagsType form and render new templates

AntePrkacin committed:
NGSTACK-977 add 'autocomplete_provide_hidden_tags' parameter to the configTreeBuilder and contextualize it

AntePrkacin committed:
NGSTACK-977 modify autoCompleteAction() method in FieldController to look at 'autocomplete_provide_hidden_tags' config parameter

pspanja commented:
@emodric depends a bit on how we solve some other open questions, but we should probably take care that non-visible tags are not indexed.

AntePrkacin committed:
NGSTACK-977 make autocomplete suggestions not show invisible tags when searching to add a tag to some content

AntePrkacin committed:
NGSTACK-977 add data about tag visibility when searching for tags in Netgen Tags adminUI

AntePrkacin committed:
NGSTACK-977 fix tag converting to synonym process and merging process and add the ability to select hidden tags in those processes

AntePrkacin committed:
NGSTACK-977 remove dump statements from 'hide' and 'reveal' twig templates

AntePrkacin committed:
NGSTACK-977 update 'is_hidden' check to 'is_invisible' check in methods that have showHiddenTags as parameter in DoctrineDatabase

AntePrkacin committed:
NGSTACK-977 fix revealing a hidden synonym by checking the 'main_tag_id' property

AntePrkacin committed:
NGSTACK-977 update testConvertToSynonym to check 'is_invisible' property

AntePrkacin commented:
https://github.com/netgen/TagsBundle/pull/169/commits/fd4e11fb902dad1e17e091e0328a145bbfc899dc
When converting a certain tag to a synonym, the user needs to select the tag for which they want this synonym to be (to represent). In this case, I've added the ability that the user can select a hidden tag for which they want to create a synonym. If the tag they (the user) chose for conversion was hidden, it will remain hidden also after the conversion. 

The same is valid when merging a tag - the user can select a hidden tag to which they want to merge some tag.

AntePrkacin committed:
NGSTACK-977 add some missing translations for french langauge

AntePrkacin committed:
NGSTACK-977 change policies to only have one for hiding and revealing

pspanja commented on bundle/Controller/TagViewController.php:
```diff
@@ -14,6 +14,10 @@ final class TagViewController extends Controller
      */
     public function viewAction(TagView $view): TagView
     {
+        if ($view->getTag()->isInvisible) {
```
@emodric should this include the check for the `show_hidden_tags` configuration option?
